### ISSUE TITLE
Would it be possible to introduce a `@param-covariant` annotation

### ISSUE THREAD

**Issue by VincentLanglet** [vimeo/psalm#6042]
It's better to explain it with an exemple: https://psalm.dev/r/5606351dd1

When changing a `@template` to `@template-covariant`, some methods are throwing an error:
```
Template param T of X is marked covariant and cannot be used here
```
 some other methods dont.

The purpose of @param-covariant is to say that we will only use methods compatible with the `template-covariant` annotation, so the invariant generic can be considered as covariant for this method.

From the example of https://psalm.dev/docs/annotating_code/templated_annotations/
> Here we're not doing anything bad – we're just iterating over an array of objects. But Psalm still gives that same basic error – "getNoises expects a Collection<Animal>, but Collection<Dog> was passed".

This could now be fixed to
```
/**
 * @param Collection<Animal> $collection
 * @param-covariant $collection
 */
function getNoises(Collection $collection) : void {
    foreach ($collection->list as $animal) {
        echo $animal->getNoise();
    }
}
```

**Comment by muglug** [comment]
Is there precedent in other languages like Java, C# or Hack? In the past this has very strongly guided my accept/reject inclination for such suggestions.

**Comment by VincentLanglet** [comment]
> Is there precedent in other languages like Java, C# or Hack? In the past this has very strongly guided my accept/reject inclination for such suggestions.

I don't know other languages so I can't be sure, but I made some research on it, and find nothing about this.

On the other, I think the `@param-covariant` is already implemented in psalm, just in a less-intuitive way.
https://psalm.dev/r/ff52db46e5

When using 
```
 * @template T of object
 * @param X<T> $x
```
I can pass X<object> as long as stricter type like X<stdClass>.
In the method implementation I cannot use method related to the template (like `setObject`) here, because it would expect to be called with `T`.

Taking the exemple of the variance, 
- the getNoise method can be fixed https://psalm.dev/r/79f9763586
- if the addAnimal was fixed the same way, it will still return an error https://psalm.dev/r/1cbb3b929b

If you consider this is the right way to represent what could be called a `@param-covariant` annotation, maybe the doc could be updated ? WDYT about this "template-hack" @muglug ?
